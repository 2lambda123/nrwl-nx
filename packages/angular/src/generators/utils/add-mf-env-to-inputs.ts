import { type Tree, readNxJson, updateNxJson } from '@nx/devkit';

export function addMfEnvToTargetDefaultInputs(tree: Tree) {
  const nxJson = readNxJson(tree);
  const webpackExecutor = '@nx/angular:webpack-browser';
  const mfEnvVar = 'NX_MF_DEV_SERVER_STATIC_REMOTES';

  let mfEnvVarExists = false;
  if (nxJson.targetDefaults && webpackExecutor in nxJson.targetDefaults) {
    const webpackExecutorTargetDefaults =
      nxJson.targetDefaults[webpackExecutor];

    if (webpackExecutorTargetDefaults.inputs) {
      for (const webpackExecutorTargetDefaultsInput of webpackExecutorTargetDefaults.inputs) {
        if (typeof webpackExecutorTargetDefaultsInput === 'object') {
          const [key, value] = Object.entries(
            webpackExecutorTargetDefaultsInput
          )[0];

          if (key === 'env' && value === mfEnvVar) {
            mfEnvVarExists = true;
            break;
          }
        }
      }
    } else {
      nxJson.targetDefaults[webpackExecutor].inputs = [];
    }
  } else if (!nxJson.targetDefaults[webpackExecutor]) {
    nxJson.targetDefaults[webpackExecutor] = {
      inputs: [],
    };
  } else if (!nxJson.targetDefaults) {
    nxJson.targetDefaults = {
      [webpackExecutor]: {
        inputs: [],
      },
    };
  }

  if (!mfEnvVarExists) {
    nxJson.targetDefaults[webpackExecutor].inputs.push({ env: mfEnvVar });
  }

  updateNxJson(tree, nxJson);
}
