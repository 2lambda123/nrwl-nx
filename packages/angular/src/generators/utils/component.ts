import {
  joinPathFragments,
  names,
  normalizePath,
  readProjectConfiguration,
  readWorkspaceConfiguration,
  Tree,
} from '@nrwl/devkit';

export type ComponentGenerationOptions = {
  name: string;
  flat?: boolean;
  path?: string;
  project?: string;
  type?: string;
};
export type ComponentFileInfo = {
  componentDirectory: string;
  componentFileName: string;
  componentFilePath: string;
};

export function getComponentFileInfo(
  tree: Tree,
  options: ComponentGenerationOptions
): ComponentFileInfo {
  const project =
    options.project ?? readWorkspaceConfiguration(tree).defaultProject;
  const { root, sourceRoot, projectType } = readProjectConfiguration(
    tree,
    project
  );
  const { fileName: normalizedName } = names(options.name);

  const componentFileName = `${normalizedName}.${
    options.type ? names(options.type).fileName : 'component'
  }`;

  const projectSourceRoot = sourceRoot ?? joinPathFragments(root, 'src');
  const path =
    options.path ??
    joinPathFragments(
      projectSourceRoot,
      projectType === 'application' ? 'app' : 'lib'
    );
  const componentDirectory = options.flat
    ? normalizePath(path)
    : joinPathFragments(path, normalizedName);

  const componentFilePath = joinPathFragments(
    componentDirectory,
    `${componentFileName}.ts`
  );

  return { componentDirectory, componentFileName, componentFilePath };
}
