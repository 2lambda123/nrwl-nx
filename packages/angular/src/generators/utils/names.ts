import { normalizePath } from '@nrwl/devkit';

export type NameInfo = { name: string; path: string };

export function parseName(rawName: string): NameInfo {
  const parsedName = normalizePath(rawName).split('/');
  const name = parsedName.pop();
  const path = parsedName.join('/');

  return { name, path };
}
