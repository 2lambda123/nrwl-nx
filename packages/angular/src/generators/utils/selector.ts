import type { Tree } from '@nrwl/devkit';
import { names, readNxJson } from '@nrwl/devkit';

export function buildSelector(
  tree: Tree,
  name: string,
  prefix: string | undefined,
  projectPrefix: string | undefined,
  casing: keyof Pick<ReturnType<typeof names>, 'fileName' | 'propertyName'>
): string {
  let selector = name;
  prefix ??= projectPrefix ?? readNxJson(tree)?.npmScope;
  if (prefix) {
    selector = `${prefix}-${selector}`;
  }

  return names(selector)[casing];
}
