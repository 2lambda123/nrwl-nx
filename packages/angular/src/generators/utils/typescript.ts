import { Tree } from '@nrwl/devkit';

export function getRootTsConfigPath(tree: Tree): string {
  for (const path of ['tsconfig.base.json', 'tsconfig.json']) {
    if (tree.exists(path)) {
      return path;
    }
  }

  throw new Error(
    'Could not find root tsconfig file. Tried with "tsconfig.base.json" and "tsconfig.json".'
  );
}
