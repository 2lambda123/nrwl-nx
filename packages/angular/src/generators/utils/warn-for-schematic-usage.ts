export function warnForSchematicUsage<T>(convertedGenerator: T): T {
  console.warn(
    'Running generators as schematics is deprecated and will be removed in v17.'
  );

  return convertedGenerator;
}
