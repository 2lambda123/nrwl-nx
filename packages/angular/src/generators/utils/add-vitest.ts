import {
  addDependenciesToPackageJson,
  ensurePackage,
  joinPathFragments,
  readJson,
  type Tree,
  updateJson,
} from '@nx/devkit';
import { analogVitestAngular, nxVersion } from '../../utils/versions';

export type AddVitestOptions = {
  name: string;
  projectRoot: string;
  skipPackageJson: boolean;
  strict: boolean;
};

export async function addVitest(
  tree: Tree,
  options: AddVitestOptions
): Promise<void> {
  if (!options.skipPackageJson) {
    process.env.npm_config_legacy_peer_deps ??= 'true';

    addDependenciesToPackageJson(
      tree,
      {},
      {
        '@analogjs/vitest-angular': analogVitestAngular,
        '@analogjs/vite-plugin-angular': analogVitestAngular,
      },
      undefined,
      true
    );
  }

  const { createOrEditViteConfig, viteConfigurationGenerator } = ensurePackage<
    typeof import('@nx/vite')
  >('@nx/vite', nxVersion);

  const relativeTestSetupPath = joinPathFragments('src', 'test-setup.ts');
  const isEsmImplicit = readJson(tree, 'package.json').type === 'module';

  createOrEditViteConfig(
    tree,
    {
      project: options.name,
      includeLib: false,
      includeVitest: true,
      inSourceTests: false,
      imports: [`import angular from '@analogjs/vite-plugin-angular'`],
      plugins: ['angular()'],
      setupFile: relativeTestSetupPath,
      useEsmExtension: !isEsmImplicit,
    },
    true
  );

  const setupFile = joinPathFragments(
    options.projectRoot,
    relativeTestSetupPath
  );
  if (!tree.exists(setupFile)) {
    tree.write(
      setupFile,
      `import '@analogjs/vitest-angular/setup-zone';

import {
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting,
} from '@angular/platform-browser-dynamic/testing';
import { getTestBed } from '@angular/core/testing';

getTestBed().initTestEnvironment(
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting()
);
`
    );
  }

  await viteConfigurationGenerator(tree, {
    project: options.name,
    newProject: true,
    uiFramework: 'angular',
    includeVitest: true,
    testEnvironment: 'jsdom',
  });

  updateJson(
    tree,
    joinPathFragments(options.projectRoot, 'tsconfig.spec.json'),
    (tsConfig) => {
      tsConfig.files = [...(tsConfig.files ?? []), relativeTestSetupPath];
      return tsConfig;
    }
  );

  updateJson(
    tree,
    joinPathFragments(options.projectRoot, 'tsconfig.app.json'),
    (tsConfig) => {
      tsConfig.exclude = [...(tsConfig.exclude ?? []), relativeTestSetupPath];
      return tsConfig;
    }
  );
}
