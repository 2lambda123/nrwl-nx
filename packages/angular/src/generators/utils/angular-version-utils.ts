import type { Tree } from '@nrwl/devkit';
import { readJson } from '@nrwl/devkit';
import { clean, coerce, major } from 'semver';
import { angularVersion } from '../../utils/versions';

export function getGeneratorDirectoryForInstalledAngularVersion(
  tree: Tree
): string | null {
  const majorAngularVersion = getInstalledAngularMajorVersion(tree);

  const directoryDictionary = {
    14: 'angular-v14',
  };

  return directoryDictionary[majorAngularVersion] ?? null;
}

export function getInstalledAngularVersion(tree: Tree): string {
  const pkgJson = readJson(tree, 'package.json');
  const installedAngularVersion =
    pkgJson.dependencies && pkgJson.dependencies['@angular/core'];

  if (
    !installedAngularVersion ||
    installedAngularVersion === 'latest' ||
    installedAngularVersion === 'next'
  ) {
    return clean(angularVersion) ?? coerce(angularVersion).version;
  }

  return (
    clean(installedAngularVersion) ?? coerce(installedAngularVersion).version
  );
}

export function getInstalledAngularMajorVersion(tree: Tree): number {
  return major(getInstalledAngularVersion(tree));
}

export function getInstalledAngularVersionInfo(tree: Tree) {
  const installedVersion = getInstalledAngularVersion(tree);

  return {
    version: installedVersion,
    major: major(installedVersion),
  };
}

export function getInstalledPackageVersionInfo(tree: Tree, pkgName: string) {
  try {
    const version =
      readJson(tree, 'package.json').dependencies?.[pkgName] ??
      readJson(tree, 'package.json').devDependencies?.[pkgName];

    return { major: major(coerce(version)), version };
  } catch {
    return null;
  }
}
