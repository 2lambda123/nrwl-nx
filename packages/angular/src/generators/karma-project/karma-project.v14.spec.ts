import type { Tree } from '@nrwl/devkit';
import * as devkit from '@nrwl/devkit';
import {
  readProjectConfiguration,
  updateJson,
  updateProjectConfiguration,
} from '@nrwl/devkit';
import { createTreeWithEmptyV1Workspace } from '@nrwl/devkit/testing';
import { karmaProjectGenerator } from './karma-project';
import libraryGenerator from '../library/library';
import { Linter } from '@nrwl/linter';
import { UnitTestRunner } from '../../utils/test-runners';
import applicationGenerator from '../application/application';

describe('karmaProject', () => {
  let tree: Tree;
  describe('support angular v14', () => {
    beforeEach(async () => {
      tree = createTreeWithEmptyV1Workspace();

      updateJson(tree, 'package.json', (json) => ({
        ...json,
        dependencies: {
          ...json.dependencies,
          '@angular/core': '14.1.0',
        },
      }));

      await libraryGenerator(tree, {
        name: 'lib1',
        buildable: false,
        linter: Linter.EsLint,
        publishable: false,
        simpleModuleName: false,
        skipFormat: false,
        unitTestRunner: UnitTestRunner.None,
      });

      await applicationGenerator(tree, {
        name: 'app1',
        unitTestRunner: UnitTestRunner.None,
      });
    });

    it('should throw when there is already a test target', async () => {
      devkit.updateJson(tree, 'workspace.json', (json) => {
        json.projects['lib1'].architect.test = {};
        return json;
      });

      await expect(
        karmaProjectGenerator(tree, { project: 'lib1' })
      ).rejects.toThrow('"lib1" already has a test target.');
    });

    it('should generate files', async () => {
      expect(tree.exists('karma.conf.js')).toBeFalsy();
      await karmaProjectGenerator(tree, { project: 'lib1' });

      expect(tree.exists('/libs/lib1/karma.conf.js')).toBeTruthy();
      expect(tree.exists('/libs/lib1/tsconfig.spec.json')).toBeTruthy();
      expect(
        tree.read('/libs/lib1/tsconfig.spec.json', 'utf-8')
      ).toMatchSnapshot();
      expect(tree.exists('karma.conf.js')).toBeTruthy();
      expect(
        tree.read('/libs/lib1/project.json', 'utf-8')
      ).toMatchInlineSnapshot(`null`);
    });

    it('should create a karma.conf.js', async () => {
      await karmaProjectGenerator(tree, { project: 'lib1' });

      const karmaConf = tree.read('libs/lib1/karma.conf.js').toString();
      expect(karmaConf).toMatchSnapshot();
    });

    it('should update the project tsconfig.json to reference the tsconfig.spec.json', async () => {
      await karmaProjectGenerator(tree, { project: 'lib1' });

      const tsConfig = devkit.readJson(tree, 'libs/lib1/tsconfig.json');
      expect(tsConfig.references).toContainEqual({
        path: './tsconfig.spec.json',
      });
    });

    it('should format files', async () => {
      jest.spyOn(devkit, 'formatFiles');

      await karmaProjectGenerator(tree, { project: 'lib1' });

      expect(devkit.formatFiles).toHaveBeenCalled();
    });

    describe('library', () => {
      it('should update the workspace config correctly', async () => {
        await karmaProjectGenerator(tree, { project: 'lib1' });

        const workspaceJson = devkit.readJson(tree, 'workspace.json');
        expect(workspaceJson.projects.lib1.architect.test).toEqual({
          builder: '@angular-devkit/build-angular:karma',
          options: {
            tsConfig: 'libs/lib1/tsconfig.spec.json',
            karmaConfig: 'libs/lib1/karma.conf.js',
            main: 'libs/lib1/src/test.ts',
          },
        });
      });

      it('should create a tsconfig.spec.json', async () => {
        await karmaProjectGenerator(tree, { project: 'lib1' });

        const tsConfig = devkit.readJson(tree, 'libs/lib1/tsconfig.spec.json');
        expect(tsConfig).toEqual({
          extends: './tsconfig.json',
          compilerOptions: {
            outDir: '../../dist/out-tsc',
            types: ['jasmine', 'node'],
          },
          files: ['src/test.ts'],
          include: ['**/*.spec.ts', '**/*.test.ts', '**/*.d.ts'],
        });
      });
    });

    describe('applications', () => {
      it('should update the workspace config correctly', async () => {
        await karmaProjectGenerator(tree, { project: 'app1' });

        const workspaceJson = devkit.readJson(tree, 'workspace.json');
        expect(workspaceJson.projects.app1.architect.test).toEqual({
          builder: '@angular-devkit/build-angular:karma',
          options: {
            main: 'apps/app1/src/test.ts',
            polyfills: 'apps/app1/src/polyfills.ts',
            tsConfig: 'apps/app1/tsconfig.spec.json',
            karmaConfig: 'apps/app1/karma.conf.js',
            styles: [],
            scripts: [],
            assets: [],
          },
        });
      });

      it('should create a tsconfig.spec.json', async () => {
        await karmaProjectGenerator(tree, { project: 'app1' });

        const tsConfig = devkit.readJson(tree, 'apps/app1/tsconfig.spec.json');
        expect(tsConfig).toEqual({
          extends: './tsconfig.json',
          compilerOptions: {
            outDir: '../../dist/out-tsc',
            types: ['jasmine', 'node'],
          },
          files: ['src/test.ts'],
          include: ['**/*.spec.ts', '**/*.test.ts', '**/*.d.ts'],
        });
      });

      it('should update the workspace config correctly when using old style ng polyfills', async () => {
        tree.write('apps/app1/src/polyfills.ts', 'import zone.js;');
        const project = readProjectConfiguration(tree, 'app1');
        project.targets.build.options.polyfills = 'apps/app1/src/polyfills.ts';
        updateProjectConfiguration(tree, 'app1', project);

        await karmaProjectGenerator(tree, { project: 'app1' });

        const workspaceJson = devkit.readJson(tree, 'workspace.json');
        expect(workspaceJson.projects.app1.architect.test).toEqual({
          builder: '@angular-devkit/build-angular:karma',
          options: {
            polyfills: 'apps/app1/src/polyfills.ts',
            main: 'apps/app1/src/test.ts',
            tsConfig: 'apps/app1/tsconfig.spec.json',
            karmaConfig: 'apps/app1/karma.conf.js',
            styles: [],
            scripts: [],
            assets: [],
          },
        });
      });

      it('should create a tsconfig.spec.json  when using old style ng polyfills', async () => {
        tree.write('apps/app1/src/polyfills.ts', 'import zone.js;');
        const project = readProjectConfiguration(tree, 'app1');
        project.targets.build.options.polyfills = 'apps/app1/src/polyfills.ts';
        updateProjectConfiguration(tree, 'app1', project);

        await karmaProjectGenerator(tree, { project: 'app1' });

        const tsConfig = devkit.readJson(tree, 'apps/app1/tsconfig.spec.json');
        expect(tsConfig).toEqual({
          extends: './tsconfig.json',
          compilerOptions: {
            outDir: '../../dist/out-tsc',
            types: ['jasmine', 'node'],
          },
          files: ['src/test.ts', 'src/polyfills.ts'],
          include: ['**/*.spec.ts', '**/*.test.ts', '**/*.d.ts'],
        });
      });
    });

    describe('--root-project', () => {
      it('should support a project located at the root', async () => {
        await applicationGenerator(tree, {
          name: 'root-app',
          unitTestRunner: UnitTestRunner.None,
          rootProject: true,
        });
        await karmaProjectGenerator(tree, { project: 'root-app' });

        expect(tree.exists('karma.conf.js')).toBe(true);
        expect(tree.read('karma.conf.js', 'utf-8')).toMatchSnapshot();
        expect(tree.exists('tsconfig.spec.json')).toBe(true);
        const { references } = devkit.readJson(tree, 'tsconfig.json');
        expect(references).toContainEqual({
          path: './tsconfig.spec.json',
        });
        const project = devkit.readProjectConfiguration(tree, 'root-app');
        expect(project.targets.test.options).toStrictEqual({
          polyfills: 'src/polyfills.ts',
          main: 'src/test.ts',
          tsConfig: 'tsconfig.spec.json',
          karmaConfig: 'karma.conf.js',
          styles: [],
          scripts: [],
          assets: [],
        });
      });

      it('should generate the right karma.conf.js file for a nested project in a workspace with a project at the root', async () => {
        await applicationGenerator(tree, {
          name: 'root-app',
          unitTestRunner: UnitTestRunner.Karma,
          rootProject: true,
        });
        await libraryGenerator(tree, {
          name: 'nested-lib',
          unitTestRunner: UnitTestRunner.None,
        });
        await karmaProjectGenerator(tree, { project: 'nested-lib' });

        expect(tree.exists('libs/nested-lib/karma.conf.js')).toBe(true);
        expect(
          tree.read('libs/nested-lib/karma.conf.js', 'utf-8')
        ).toMatchSnapshot();
      });
    });
  });
});
