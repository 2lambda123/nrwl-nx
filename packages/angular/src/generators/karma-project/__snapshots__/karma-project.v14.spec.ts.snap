// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`karmaProject support angular v14 --root-project should generate the right karma.conf.js file for a nested project in a workspace with a project at the root 1`] = `
"// Karma configuration file, see link for more information
// https://karma-runner.github.io/6.4/config/configuration-file.html

const { join } = require('path');
const setBaseKarmaConfig = require('../../karma.conf');

module.exports = function (config) {
  setBaseKarmaConfig(config);

  config.set({
    coverageReporter: {
      dir: join(__dirname, '../../coverage/libs/nested-lib')
    }
  });
};
"
`;

exports[`karmaProject support angular v14 --root-project should support a project located at the root 1`] = `
"// Karma configuration file, see link for more information
// https://karma-runner.github.io/6.4/config/configuration-file.html

const { constants } = require('karma');
const { join } = require('path');

module.exports = (config) => {
  config.set({
    basePath: '',
    frameworks: ['jasmine', '@angular-devkit/build-angular'],
    plugins: [
      require('karma-jasmine'),
      require('karma-chrome-launcher'),
      require('karma-jasmine-html-reporter'),
      require('karma-coverage'),
      require('@angular-devkit/build-angular/plugins/karma'),
    ],
    client: {
      jasmine: {
        // you can add configuration options for Jasmine here
        // the possible options are listed at https://jasmine.github.io/api/edge/Configuration.html
        // for example, you can disable the random execution with \`random: false\`
        // or set a specific seed with \`seed: 4321\`
      },
      clearContext: false, // leave Jasmine Spec Runner output visible in browser
    },
    jasmineHtmlReporter: {
      suppressAll: true, // removes the duplicated traces
    },
    coverageReporter: {
      dir: join(__dirname, 'coverage/root-app'),
      subdir: '.',
      reporters: [{ type: 'html' }, { type: 'text-summary' }],
    },
    reporters: ['progress', 'kjhtml'],
    port: 9876,
    colors: true,
    logLevel: constants.LOG_INFO,
    autoWatch: true,
    browsers: ['Chrome'],
    singleRun: true,
    restartOnFileChange: true,
  });
};
"
`;

exports[`karmaProject support angular v14 should create a karma.conf.js 1`] = `
"// Karma configuration file, see link for more information
// https://karma-runner.github.io/6.4/config/configuration-file.html

const { join } = require('path');
const getBaseKarmaConfig = require('../../karma.conf');

module.exports = function(config) {
  const baseConfig = getBaseKarmaConfig();
  config.set({
    ...baseConfig,
    coverageReporter: {
      ...baseConfig.coverageReporter,
      dir: join(__dirname, '../../coverage/libs/lib1')
    }
  });
};
"
`;

exports[`karmaProject support angular v14 should generate files 1`] = `
"{
  \\"extends\\": \\"./tsconfig.json\\",
  \\"compilerOptions\\": {
    \\"outDir\\": \\"../../dist/out-tsc\\",
    \\"types\\": [\\"jasmine\\", \\"node\\"]
  },
  \\"files\\": [\\"src/test.ts\\"],
  \\"include\\": [\\"**/*.spec.ts\\", \\"**/*.test.ts\\", \\"**/*.d.ts\\"]
}
"
`;
